packages/animations/lib/src/modal.dart

```diff
@@ -0,0 +1,216 @@
+  _ModalRoute({
+    this.barrierColor,
+    this.barrierDismissible = true,
+    this.barrierLabel,
+    this.transitionDuration,
+    this.reverseTransitionDuration,
+    _ModalTransitionBuilder transitionBuilder,
+    @required this.builder,
+  })  : assert(barrierDismissible != null),
+        assert(!barrierDismissible || barrierLabel != null),
+        _transitionBuilder = transitionBuilder;
+
+  @override
+  final Color barrierColor;
+
+  @override
+  final bool barrierDismissible;
+
+  @override
+  final String barrierLabel;
+
+  @override
+  final Duration transitionDuration;
+
+  @override
+  final Duration reverseTransitionDuration;
+
+  /// The primary contents of the modal.
+  final WidgetBuilder builder;
+
+  final _ModalTransitionBuilder _transitionBuilder;
+
+  @override
+  Widget buildPage(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+  ) {
+    final ThemeData theme = Theme.of(context);
+    return Semantics(
+      child: SafeArea(
+        child: Builder(
+          builder: (BuildContext context) {
+            final Widget child = Builder(builder: builder);
+            return theme != null ? Theme(data: theme, child: child) : child;
+          },
+        ),
+      ),
+      scopesRoute: true,
+      explicitChildNodes: true,
+    );
+  }
+
+  @override
+  Widget buildTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    return _transitionBuilder(
+      context,
+      animation,
+      secondaryAnimation,
+      child,
+    );
+  }
+}
+
+/// A configuration object containing the properties needed to implement a
+/// modal route.
+///
+/// The `barrierDismissible` argument is used to determine whether this route
+/// can be dismissed by tapping the modal barrier. This argument defaults
+/// to true. If `barrierDismissible` is true, a non-null `barrierLabel` must be
+/// provided.
+///
+/// The `barrierLabel` argument is the semantic label used for a dismissible
+/// barrier. This argument defaults to "Dismiss".
+abstract class ModalConfiguration {
+  /// Creates a modal configuration object that provides the necessary
+  /// properties to implement a modal route.
+  ///
+  /// [barrierDismissible] configures whether or not tapping the modal's
+  /// scrim dismisses the modal. [barrierLabel] sets the semantic label for
+  /// a dismissible barrier. [barrierDismissible] cannot be null. If
+  /// [barrierDismissible] is true, the [barrierLabel] cannot be null.
+  ///
+  /// [transitionDuration] and [reverseTransitionDuration] determine the
+  /// duration of the transitions when the modal enters and exits the
+  /// application. [transitionDuration] and [reverseTransitionDuration]
+  /// cannot be null.
+  ModalConfiguration({
+    this.barrierColor,
+    @required this.barrierDismissible,
+    this.barrierLabel,
+    this.transitionDuration,
```

Comment:
We should decide wether we want these to be supplied via overrides or by passing them in in the constructor. If we take them via constructor, then we should assert that they have the right value here (and have subclasses pas them in in the super constructor call). If we want them to be provided via override, we should remove them from the constructor and just have abstract `Duration get transitionDuration` getters in the class.

I am voting to do the latter.